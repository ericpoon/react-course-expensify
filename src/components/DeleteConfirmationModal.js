import React from 'react';
import Modal from 'react-modal';
import moment from 'moment/moment';
import numeral from 'numeral';

export default class DeleteConfirmationModal extends React.Component {
    render() {
        const {expense, isOpen, onRemoveConfirm, onRemoveCancel} = this.props;
        return (
            <Modal
                isOpen={isOpen}
                onRequestClose={onRemoveCancel}
                className={'modal'}
                portalClassName={'modal__wrapper'}
                closeTimeoutMS={200}
                appElement={document.getElementById('app')}
            >
                <div className={'modal__body'}>
                    <h3>Removing an expense</h3>
                    <p>Are you sure you want to remove the expense for <span>{expense.description}</span>?
                    </p>
                    <p>It's {numeral(expense.amount).format('$0,0.00')} and spent
                        on {moment(expense.createdAt).format('MMM D, YYYY')}.</p>
                </div>
                <div className={'modal__actions'}>
                    <button onClick={onRemoveConfirm} className={'button button--danger'}>Confirm</button>
                    <button onClick={onRemoveCancel} className={'button button--secondary'}>Cancel</button>
                </div>
            </Modal>
        );
    }
}