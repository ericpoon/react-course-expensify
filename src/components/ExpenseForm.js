import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        const {expense} = this.props;
        const description = expense ? expense.description : '';
        const note = expense ? expense.note : '';
        const amount = expense ? expense.amount.toString() : '';
        const createdAt = expense ? moment(expense.createdAt) : moment();

        this.state = {
            description,
            note,
            amount,
            createdAt,
            calendarFocused: false,
            error: '',
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState({description});
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState({note});
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}\.?\d{0,2}$/)) {
            this.setState({amount});
        }
    };

    onDateChange = (date) => {
        if (date) this.setState({createdAt: date});
    };

    onFocusChange = ({focused}) => {
        this.setState({calendarFocused: focused});
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (this.state.description && this.state.amount) {
            this.setState({error: ''});
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            });
        } else {
            this.setState({error: 'Please provide description and amount'});
        }
    };

    render() {
        return (
            <form onSubmit={this.onSubmit} className={'form'}>
                {this.state.error && <p className={'form__error'}>{this.state.error}</p>}
                <input
                    type={'text'}
                    placeholder={'Description'}
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    className={'text-input'}
                />
                <input
                    type={'text'}
                    placeholder={'Amount'}
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    className={'text-input'}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    // This extends the date picker to take the full width
                    block
                />
                <textarea
                    placeholder={'Add a note for your expense (optional)'}
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    className={'textarea'}
                />
                <div>
                    {/*This div prevents the button from taking the full width*/}
                    <button className={'button'}>{this.props.submitButtonValue}</button>
                </div>
            </form>
        );
    }
}

export default ExpenseForm;