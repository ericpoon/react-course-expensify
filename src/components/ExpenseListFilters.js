import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState({calendarFocused});
    };

    onSortChange = (e) => {
        switch (e.target.value) {
        case 'date':
            this.props.sortByDate();
            break;
        case 'amount':
            this.props.sortByAmount();
            break;
        }
    };

    render() {
        return (
            <div className={'content-container'}>
                <div className={'input-group'}>
                    <div className={'input-group__item'}>
                        <input
                            type={'text'}
                            value={this.props.filters.text}
                            placeholder={'Search expenses'}
                            onChange={this.onTextChange}
                            className={'text-input'}
                        />
                    </div>
                    <div className={'input-group__item'}>
                        <select
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}
                            className={'select'}
                        >
                            <option value={'date'}>Date</option>
                            <option value={'amount'}>Amount</option>
                        </select>
                    </div>
                    <div className={'input-group__item'}>
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            startDateId={'start'}
                            endDateId={'end'}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);