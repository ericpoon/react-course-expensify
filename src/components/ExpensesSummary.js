import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { selectExpensesTotal, selectFilteredExpenses } from '../selectors/expenses';

export const ExpensesSummary = (props) => {
  const { expenseCount: count, hiddenCount } = props;
  const total = numeral(props.expenseTotal).format('$0,0.00');
  const expenseWord = count === 1 ? 'expense' : 'expenses';
  const hiddenExpenseWord = hiddenCount === 1 ? 'expense' : 'expenses';
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{count}</span> {expenseWord}, totalling <span>{total}</span>
        </h1>
        {hiddenCount > 0
        && <p>{hiddenCount} {hiddenExpenseWord} hidden, clear filters to view all.</p>}
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const totalCountWithoutFilters = state.expenses.length;
  const expenses = selectFilteredExpenses(state.expenses, state.filters);
  return {
    hiddenCount: totalCountWithoutFilters - expenses.length,
    expenseCount: expenses.length,
    expenseTotal: selectExpensesTotal(expenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
