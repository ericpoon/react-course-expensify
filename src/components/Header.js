import React from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth';

export const Header = (props) => (
    <header className={'header'}>
        <div className={'content-container'}>
            <div className={'header__content'}>
                {/* Need to pass exact={'true'} to suppress warning */}
                <Link to={'/dashboard'} exact className={'header__title'}>
                    <h1>Expensify</h1>
                </Link>
                <div className={'header__actions'}>
                    <button
                        onClick={() => null}
                        className={'button button--no-background'}
                    >
                        {props.userName}
                    </button>

                    <button
                        onClick={props.startLogout}
                        className={'button button--no-background'}
                    >
                        Logout
                    </button>
                </div>
            </div>
        </div>
    </header>
);

const mapStateToProps = (state) => ({
    userName: state.auth.name,
});

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);