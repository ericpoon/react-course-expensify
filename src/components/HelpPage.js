import React from 'react';

export default () => (
    <div>This my help component.</div>
);