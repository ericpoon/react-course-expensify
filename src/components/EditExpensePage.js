import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import DeleteConfirmationModal from './DeleteConfirmationModal';
import {startEditExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor() {
        super();
        this.state = {
            isModalOpen: false,
        };
    }

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/dashboard');
    };
    onRemoveClick = () => {
        this.setState({isModalOpen: true});
    };
    onRemoveConfirm = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/dashboard');
    };
    onRemoveCancel = () => {
        this.setState({isModalOpen: false});
    };

    render() {
        return (
            <div>
                <DeleteConfirmationModal
                    expense={this.props.expense}
                    isOpen={this.state.isModalOpen}
                    onRemoveConfirm={this.onRemoveConfirm}
                    onRemoveCancel={this.onRemoveCancel}
                />
                <div className={'page-header'}>
                    <div className={'content-container'}>
                        <h1 className={'page-header__title'}>Edit Expense</h1>
                    </div>
                </div>
                <div className={'content-container'}>
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                        submitButtonValue={'Update Expense'}
                    />
                    <button
                        onClick={this.onRemoveClick}
                        className={'button button--secondary'}
                    >
                        Remove Expense
                    </button>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    const {id} = props.match.params;
    return {
        expense: state.expenses.find(e => e.id === id),
    };
};

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);