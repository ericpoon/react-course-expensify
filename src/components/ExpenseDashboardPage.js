import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './expensesSummary';

export default () => (
    <div>
        <ExpensesSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
);