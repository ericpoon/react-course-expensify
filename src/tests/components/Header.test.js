import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

let wrapper;

beforeEach(() => {
  wrapper = shallow(<Header />);
});

it('should render Header correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should call startLogout on button click', () => {
  const startLogout = jest.fn();
  wrapper.setProps({ startLogout });
  wrapper.find('button').at(1).simulate('click');
  expect(startLogout).toHaveBeenCalled();
});
