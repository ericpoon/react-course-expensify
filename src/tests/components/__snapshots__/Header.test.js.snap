// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        exact={true}
        replace={false}
        to="/dashboard"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <div
        className="header__actions"
      >
        <button
          className="button button--no-background"
          onClick={[Function]}
        />
        <button
          className="button button--no-background"
        >
          Logout
        </button>
      </div>
    </div>
  </div>
</header>
`;
